### ISSUE TITLE
Unimformative error in compare.py

### ISSUE THREAD

**Issue by rlorigro** [whatshap/whatshap#342]
Hi, 

I am attempting to use whatshap compare to gather some phasing stats on an HG002 assembly I have, using the GIAB HG002 VCF as aligned to hg37. However, I get an error that is difficult to interpret when running:

```
Found 1 sample(s) in input VCF
Reporting results for sample sample
+ whatshap compare --tsv-pairwise paternal_dipcall_PS.wh_stats/paternal_dipcall_PS.pairwise.tsv --tsv-multiway paternal_dipcall_PS.wh_stats/paternal_dipcall_PS.multiway.tsv --switch-error-bed paternal_dipcall_PS.wh_stats/paternal_dipcall_PS.switch_error.bed --plot-blocksizes paternal_dipcall_PS.wh_stats/paternal_dipcall_PS.blocksizes.pdf --plot-sum-of-blocksizes paternal_dipcall_PS.wh_stats/paternal_dipcall_PS.sum_of_blocksizes.pdf --longest-block-tsv paternal_dipcall_PS.wh_stats/paternal_dipcall_PS.longest_block.tsv paternal.dipcall.PS.vcf GIAB.HG002_vs_HG37.PATMAT.vcf
Reading phasing from 'paternal.dipcall.PS.vcf'
Reading phasing from 'GIAB.HG002_vs_HG37.PATMAT.vcf'
Chromosomes present in all VCFs: 1, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 2, 20, 21, 22, 3, 4, 5, 6, 7, 8, 9
Traceback (most recent call last):
  File "/usr/local/bin/whatshap", line 8, in <module>
    sys.exit(main())
  File "/usr/local/lib/python3.8/dist-packages/whatshap/__main__.py", line 83, in main
    module.main(args)
  File "/usr/local/lib/python3.8/dist-packages/whatshap/cli/compare.py", line 976, in main
    run_compare(**vars(args))
  File "/usr/local/lib/python3.8/dist-packages/whatshap/cli/compare.py", line 898, in run_compare
    ) = compare(
  File "/usr/local/lib/python3.8/dist-packages/whatshap/cli/compare.py", line 497, in compare
    errors = compare_block(phasing0, phasing1)
  File "/usr/local/lib/python3.8/dist-packages/whatshap/cli/compare.py", line 283, in compare_block
    total_hamming += hamming(phasing1[i], permutation[i])
  File "/usr/local/lib/python3.8/dist-packages/whatshap/cli/compare.py", line 111, in hamming
    assert len(s0) == len(s1)
AssertionError
```

Is there a simple explanation for what might be causing this? I can share data if that would help.

**Comment by marcelm** [comment]
I don’t see anything obviously wrong, so it would indeed be easiest if you could share the data with us!

**Comment by rlorigro** [comment]
Alright here are my two VCFs:
https://rlorigro-public-files.s3.us-west-1.amazonaws.com/gfase/paolo_ul_run6/phased_k31_double_coverage/dipcall/GIAB.HG002_vs_HG37.PATMAT.vcf
https://rlorigro-public-files.s3.us-west-1.amazonaws.com/gfase/paolo_ul_run6/phased_k31_double_coverage/dipcall/paternal.dipcall.PS.clean.vcf

The command is from a wdl, but it looked something like this:
```
# get stats
whatshap stats \
    --tsv ${OUTPUT}.stats.tsv \
    --block-list ${OUTPUT}.blocks.txt \
    --chr-lengths /root/chr_lengths \
    $QUERY \
    >${OUTPUT}.stats.txt

# get comparison
whatshap compare \
    --tsv-pairwise ${OUTPUT}.pairwise.tsv \
    --tsv-multiway ${OUTPUT}.multiway.tsv \
    --switch-error-bed ${OUTPUT}.switch_error.bed \
    --plot-blocksizes ${OUTPUT}.blocksizes.pdf \
    --plot-sum-of-blocksizes ${OUTPUT}.sum_of_blocksizes.pdf \
    --longest-block-tsv ${OUTPUT}.longest_block.tsv \
    $QUERY $TRUTH \
    >$OUTPUT.compare.txt
```

**Comment by marcelm** [comment]
Thanks! Are you sure, though, that these are the correct files? When I run `whatshap stats paternal.dipcall.PS.clean.vcf`, I get
```
whatshap.vcf.VcfNotSortedError: VCF not ordered: 1:6693976 appears before 1:6377
```
Also, the sample name in the GIAB file is `HG002`, but it is `default` in the `paternal.dipcall...` file, so also `whatshap compare` complains.

**Comment by rlorigro** [comment]
Hi, Sorry about that, there was some VCF preprocessing in the WDL workflow that I neglected to account for. I reuploaded the `paternal.dipcall.PS.clean.vcf` file and it should (hopefully) recreate the input that the workflow creates.

**Comment by marcelm** [comment]
Thanks, I was able to reproduce the problem with the new file. I’ll look into this.

**Comment by marcelm** [comment]
Some notes on this (more to myself, but feel free to comment if you have any thoughts).

The crash happens because of this variant in the `paternal` file (there are more like this):
```
1       247607030       .       G       C       30      GAP2    .       GT:AD:PS        1|.:0,1:37
```
The `GT` field is `1|.`, which `whatshap compare` handles incorrectly. That is, one allele is unknown in a phased genotype. I guess this could happen when a single read (or very few) covers a position on the reference and contains the alternative allele. Then a variant caller may decide that it doesn’t have enough data to call this a homozygous variant, but would rather call it `1/.` to be on the safe side.

But whatever the reason, WhatsHap shouldn’t crash, and we need to decide what to do. The error occurs when two phase sets are compared and the number of errors is computed. As a simple example, let’s say we have a phase set in samples A and B with just three variants:

A:
```
0|1
1|1
1|0
```
B:
```
0|1
0|1
1|0
```

Let h(x, y) be the Hamming distance between x and y. For the example, the number of errors is computed as h("011", "001") + h("110", "110") = 1 (that is, as h(left column of A, left column of B) + h(right column of A, right column of b))

I now change it so that the middle variant of A has an unknown allele:

A:
```
0|1
1|.
1|0
```
B:
```
0|1
0|1
1|0
```

I wonder whether it is correct to just do the computation as before, that is, the number of errors would be h("011", "001") + h("1.0", "110") = 2.

**Comment by marcelm** [comment]
I committed a fix for this. As I suggested above, the `.` is now compared literally to the corresponding allele in the respective other phase set. The alternative would be to ignore variants entirely that contain unknown alleles.

**Comment by tpesout** [comment]
Hi @marcelm, the context here is that we're using [dipcall](https://github.com/lh3/dipcall) to call phased variants from a diploid assembly.  I expect that this case is from a region where there was no sequence for one of the haplotypes (either not assembled or perhaps in a structural deletion), so the "unknown allele" case makes sense although I've never seen this in a VCF before.

I think that this case should probably be ignored instead of compared.. I confess I don't know how whatshap handles a mis-called variant or a mis-genotyped variant (like, truth says it's HOM but query says it's HET), but I think this error is similar to that and had assumed they would just be excluded from the switch/hamming calculation.

Is it possible for you to change handling of this so it's ignored?

**Comment by marcelm** [comment]
> I confess I don't know how whatshap handles a mis-called variant or a mis-genotyped variant (like, truth says it's HOM but query says it's HET)

When comparing two samples, `whatshap compare` only considers the variants that are heterozygous in both, so yes, these would just be ignored.

> Is it possible for you to change handling of this so it's ignored?

Sure!

**Comment by rlorigro** [comment]
Thanks so much for the fast response, we will try this out.

**Comment by marcelm** [comment]
I’ll make a new release when you have confirmed that it works.

**Comment by tpesout** [comment]
It looks like this worked, thanks Marcel!

**Comment by marcelm** [comment]
Thanks for the feedback! I have released WhatsHap 1.2.1. The Bioconda package should also be available soon.
